infra/modules/apim/policies/api_policy.xml

```diff
@@ -68,6 +22,128 @@
+                        }
+                        
+                        return clusters;   
+                    }" />
+                <!-- Add cluster configurations to cache -->
+                <cache-store-value key="@("oaClusters" + context.Deployment.Region + context.Api.Revision)" value="@((JArray)context.Variables["oaClusters"])" duration="86400" />
+            </when>
+        </choose>
+        <!-- Getting OpenAI routes configuration based on deployment name, region and api revision -->
+        <cache-lookup-value key="@(context.Request.MatchedParameters["deployment-id"] + "Routes" + context.Deployment.Region + context.Api.Revision)" variable-name="routes" />
+        <!-- If we can't find the configuration, it will be loaded -->
+        <choose>
+            <when condition="@(context.Variables.ContainsKey("routes") == false)">
+                <set-variable name="routes" value="@{
+                        string deploymentName = context.Request.MatchedParameters["deployment-id"];
+                        JArray clusters = (JArray)context.Variables["oaClusters"];
+                        JObject cluster = (JObject)clusters.FirstOrDefault(o => o["deploymentName"]?.Value<string>() == deploymentName);
+                        if(cluster == null)
+                        {
+                            //Error: No cluster matched the requested deployment name
+                            return new JArray() { new JObject()
+                                {
+                                    { "name", deploymentName },
+                                    { "location", "NA" },
+                                    { "url", "No routes found for the deployment (" + deploymentName + ") in the region (" + context.Deployment.Region + ")" }
+                                }
+                            };
+                        }
+                        JArray routes = (JArray)cluster["routes"];
+                        return routes;
+                    }" />
+                <!-- If no routes found for deployment, return bad request with content of routes variable -->
+                <choose>
+                    <when condition="@(((JArray)context.Variables["routes"]).ToString().Contains("No routes"))">
+                        <return-response>
+                            <set-status code="400" reason="No routes" />
+                            <set-body>@(((JArray)context.Variables["routes"]).ToString())</set-body>
+                        </return-response>
+                    </when>
+                </choose>
+                <!-- Add cluster configurations to cache -->
+                <cache-store-value key="@(context.Request.MatchedParameters["deployment-id"] + "Routes" + context.Deployment.Region + context.Api.Revision)" value="@((JArray)context.Variables["routes"])" duration="86400" />
```

Comment:
I can see this key name is used multiple times below. Shouldn't we create a local variable once and just refer to that variable name in the other locations where we store the cache key? Otherwise, we keep concateneting this long key value over and over in all places.

Reply:
Also, duration = 86400 will create some issues for people that are experimenting and changing the backend routes often to test new configurations, because it won't be cleaned up from the internal cache, they would need to manually remove that cache key from APIM, that is a hassle. My suggestion is to stick with 60 because that should be the maximum retry-after value coming from the OpenAI backends anyway, and after 1 minute all the JSON array will be reconstructed as specified in the policy and everything reset to default values.

Reply:
Great feedback @andredewes 
- I see the value of setting a variable to be referenced. Great suggestion.
- As for the duration=86400, the new approach doesn't rely on cache expiration to update the configurations, rather it is recommended to create new API revision -> update the configuration -> test it -> switch new revision to be current (this will take immediate effect as it will create new cache entries for new requests with ability to rollback to older revision if needed). This will promote safe deployment practices.